annotation/tests/test_tasks_services.py

```diff
@@ -1,30 +1,246 @@
-from annotation.models import File, ManualAnnotationTask
-from annotation.schemas.tasks import ManualAnnotationTaskInSchema
-from annotation.tasks.services import (
-    check_cross_annotating_pages,
-    create_annotation_task,
-    filter_tasks_db,
-    read_annotation_task,
-    read_annotation_tasks,
-    remove_additional_filters,
-    validate_files_info,
-    validate_ids_and_names,
-    validate_task_info,
-    validate_user_actions,
-    validate_users_info,
+from annotation.models import (
+    AnnotatedDoc,
+    AnnotationStatistics,
+    File,
+    ManualAnnotationTask,
 )
+from annotation.schemas.annotations import PageSchema, ParticularRevisionSchema
+from annotation.schemas.tasks import (
+    AgreementScoreComparingResult,
+    AgreementScoreServiceResponse,
+    AnnotationStatisticsInputSchema,
+    ExportTaskStatsInput,
+    ManualAnnotationTaskInSchema,
+    ResponseScore,
+    TaskMetric,
+    TaskStatusEnumSchema,
+)
+from annotation.tasks import services
+
+
+@pytest.fixture
+def mock_task_revisions():
+    mock_revision = AnnotatedDoc()
+    mock_revision.pages = {"1": ["data1"], "2": ["data2"]}
+    mock_revision.failed_validation_pages = [1, 2]
+    mock_revision.validated = [1, 2]
+    yield mock_revision
+
+
+@pytest.fixture
+def mock_metric():
+    metric = ManualAnnotationTask()
+    metric.task_from = datetime(2024, 1, 1)
+    metric.task_to = datetime(2024, 2, 1)
+    metric.agreement_metric = True
+    yield metric
+
+
+@pytest.fixture
+def mock_task():
+    task = ManualAnnotationTask(
+        id=1,
+        user_id=2,
+        job_id=10,
+        file_id=3,
+        pages={1, 2, 3},
+        is_validation=False,
+        status=None,
+    )
+    yield task
+
+
+@pytest.fixture
+def mock_stats(
+    mock_task: ManualAnnotationTask, mock_metric: ManualAnnotationTask
+):
+    stat1 = AnnotatedDoc()
+    stat1.task = mock_task
+    stat1.task_id = 1
+    stat1.created = datetime(2024, 1, 1, 12, 0, 0)
+    stat1.updated = datetime(2024, 1, 2, 12, 0, 0)
+    stat1.task.user_id = 2
+    stat1.task.file_id = 3
+    stat1.task.pages = [1, 2, 3]
+    stat1.task.status = TaskStatusEnumSchema.finished
+    stat1.task.agreement_metrics = [mock_metric]
+
+    stat2 = AnnotatedDoc()
+    stat2.task = mock_task
+    stat2.task_id = 2
+    stat2.created = datetime(2024, 1, 3, 12, 0, 0)
+    stat2.updated = datetime(2024, 1, 4, 12, 0, 0)
+    stat2.task.user_id = 2
+    stat2.task.file_id = 3
+    stat2.task.pages = [1, 2, 3]
+    stat2.task.status = TaskStatusEnumSchema.finished
+    stat2.task.agreement_metrics = [mock_metric]
+
+    stat3 = AnnotatedDoc()
+    stat3.task = mock_task
+    stat3.task_id = 3
+    stat3.created = datetime(2024, 1, 5, 12, 0, 0)
+    stat3.updated = datetime(2024, 1, 6, 12, 0, 0)
+    stat3.task.user_id = 2
+    stat3.task.file_id = 3
+    stat3.task.pages = [1, 2, 3]
+    stat3.task.status = TaskStatusEnumSchema.finished
+    stat3.task.agreement_metrics = [mock_metric]
```

Comment:
Provide this data in the constructor

Reply:
I’m encountering an issue with SQLAlchemy’s _declarative_constructor method when trying to initialize an AnnotatedDoc instance with keyword arguments. Specifically, I’m getting a TypeError indicating that created is an invalid keyword argument for the AnnotatedDoc class.

```

self = <AnnotatedDoc(revision=None, file_id=None, job_id=None)>
kwargs = {'created': datetime.datetime(2024, 1, 5, 12, 0), 'task_id': 3, 'updated': datetime.datetime(2024, 1, 6, 12, 0)}  
cls_ = <class 'annotation.models.AnnotatedDoc'>, k = 'created'

    def _declarative_constructor(self, **kwargs):
        """A simple constructor that allows initialization from kwargs.

        Sets attributes on the constructed instance using the names and
        values in ``kwargs``.

        Only keys that are present as
        attributes of the instance's class are allowed. These could be,
        for example, any mapped columns or relationships.
        """
        cls_ = type(self)
        for k in kwargs:
            if not hasattr(cls_, k):
>               raise TypeError(
                    "%r is an invalid keyword argument for %s" % (k, cls_.__name__)
                )
E               TypeError: 'created' is an invalid keyword argument for AnnotatedDoc
```

when I use this:

```
stat3 = AnnotatedDoc(
        tasks=[mock_task],
        task_id=3,
        created=datetime(2024, 1, 5, 12, 0, 0),
        updated=datetime(2024, 1, 6, 12, 0, 0),
        
    )
    
```